_posts/2019-05-26-redundant-order-by.md

```diff
@@ -0,0 +1,128 @@
+parallelism in the merging step of a distributed sort, and consumes more CPU and memory to sort the data.
+
+
+```sql
+SELECT *
+FROM some_table 
+   JOIN (SELECT * FROM another_table ORDER BY field) u 
+   ON some_table.key = u.key 
+```
+
+In this case, whether the tables involved in the join are sorted doesn't matter, since Presto is going to 
+build a hash lookup table out of one of them to execute the join operation. As in the previous example
+preserving the `ORDER BY` just causes the query to perform poorly. 
+
+When *does* `ORDER BY` matter? Since it is "guaranteed only for the `<query expression>` that immediately 
+contains the `<order by clause>`", only operations that are part of the same `<query expression>` are 
+sensitive to it. 
+
+A query expression is a block with the following structure:
+
+```
+<query expression> ::=
+  [ <with clause> ] 
+  <query expression body>
+  [ <order by clause> ] 
+  [ <result offset clause> ] 
+  [ <fetch first clause> ]
+```
+
+where `<query expression body>` devolves into one of the set operations (`UNION`, `INTERSECT`, `EXCEPT`), 
+a `SELECT` construct, `VALUES` or `TABLE` clause.
+
+The only operations that occur after an `ORDER BY` are `FETCH FIRST` (a.k.a., `LIMIT`) and `OFFSET`. So, 
+unless a subquery contains one of these two clauses, the query engine is free to remove the `ORDER BY` 
+clause without breaking the semantics dictated by the specification.
+
+Here's an example where the clause is meaningful:
+
+```sql
+SELECT *
+FROM some_table
+WHERE field = (
+    SELECT a 
+    FROM another_table 
+    ORDER BY b 
+    LIMIT 1)
+```
+
+## What's the catch?
+
+It is a common mistake for users to think the `ORDER BY` clause has a meaning in the language regardless of where it 
+appears in a query. The fact that, for implementation reasons, in some cases `ORDER BY` is significant for Presto 
+complicates matters. We often see users rely on this when formulating queries where aggregation or window functions 
+are sensitive to the order of their inputs:
+
+```sql
+SELECT array_agg(name) FROM (
+    SELECT *
+    FROM nation
+    ORDER BY name DESC
+)
+```
+
+```sql
+SELECT *, row_number() OVER ()
+FROM (
+    SELECT *
+    FROM nation
+    ORDER BY name DESC
+)
+```
+
+The Right™ way of doing this in SQL is to use the aggregation or window-specific `ORDER BY` clause. For the 
+examples above:
+
+
+```sql
+SELECT array_agg(name ORDER BY name DESC) 
+FROM nation
+```
+
+```sql
+SELECT *, row_number() OVER (ORDER BY name DESC)
+FROM nation
+```
+
+In order to ease the transition, the new behavior can be turned off globaly via the `optimizer.remove-redundant-sort` 
+configuration option or on a per-session basis via the `remove_redundant_sort` session property. 
+These options will be removed in a future version.
+
+Additionally, any time Presto detects a redundant `ORDER BY` clause, it will warn users about it:
+
```

Comment:
It might be worth mentioning eg SQL Server loudly rejects queries with redundant ORDER BY (eg in subquery and without TOP n). You can provide this as an illustration or a broader context.

Reply:
I didn't know that. Do you have an example of the error it shows?